### PR TITLE
 Initial implementation of the SVM model assembler

### PR DESCRIPTION
**Pull request by izeigerman**
Please note that the multi-class support as well as generated examples will be added in a subsequent PR.

### CODE REVIEW

File: m2cgen/assemblers/svm.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        self._gamma = model.gamma
+        if self._gamma == "auto" or self._gamma == "auto_deprecated":
+            self._gamma = 1.0 / n_features
+
+        self._output_size = 1
+        if type(model).__name__ == "SVC":
+            n_classes = len(model.n_support_)
+            if n_classes > 2:
+                self._output_size = n_classes
```
Comment (izeigerman) [inline m2cgen/assemblers/svm.py]: This will be used for multi-class in the follow-up PR.

File: m2cgen/assemblers/svm.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        negative_gamma = utils.sub(ast.NumVal(0), ast.NumVal(self._gamma))
+        kernel = None
+        for j in range(len(support_vector)):
+            sub_expr = utils.sub(ast.NumVal(support_vector[j]),
+                                 ast.FeatureRef(j))
+            pow_expr = ast.PowExpr(sub_expr, ast.NumVal(2))
+            if kernel:
+                kernel = utils.add(pow_expr, kernel)
+            else:
+                kernel = pow_expr
```
Comment (krinart) [inline m2cgen/assemblers/svm.py]: It's up to you, but I think using `utils.apply_op_to_expressions` (see my comment to `_linear_kernel` function) makes code simpler.

File: m2cgen/assemblers/svm.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    def _linear_kernel(self, support_vector):
+        kernel = None
+        for j in range(len(support_vector)):
+            mul_expr = utils.mul(ast.NumVal(support_vector[j]),
+                                 ast.FeatureRef(j))
+            if kernel:
+                kernel = utils.add(mul_expr, kernel)
+            else:
+                kernel = mul_expr
+        return kernel
```
Comment (krinart) [inline m2cgen/assemblers/svm.py]:
```python
return utils.apply_op_to_expressions(
    utils.mul(ast.NumVal(support_el), ast.FeatureRef(i)
    for i, support_el in enumerate(support_vector), 
    ast.BinNumOpType.ADD)
```
Does it make sense?

File: m2cgen/interpreters/interpreter.py [hunk 4]
```diff
@@ -89,9 +89,14 @@ class ToCodeInterpreter(BaseToCodeInterpreter):
     about AST.
     """
 
+    exponent_function_name = NotImplemented
+    power_function_name = NotImplemented
+    tanh_function_name = NotImplemented
```
Comment (krinart) [inline m2cgen/interpreters/interpreter.py]: `NotImplemented` is an alias to the `None`, and while in the code it looks good, in runtime we would get a native code which attempts to call function called `None`.

The best thing I can think of is just adding an assert to verify that the corresponding function is not empty.

Comment (izeigerman) [reply m2cgen/interpreters/interpreter.py]: Will do. I just followed the same logic as in https://github.com/BayesWitnesses/m2cgen/blob/master/m2cgen/interpreters/code_generator.py#L26-L34

Comment (izeigerman) [reply m2cgen/interpreters/interpreter.py]: Fixed

File: m2cgen/assemblers/svm.py (hunk 3 above)
Comment (izeigerman) [reply m2cgen/assemblers/svm.py]: Fixed

File: m2cgen/assemblers/svm.py (hunk 2 above)
Comment (izeigerman) [reply m2cgen/assemblers/svm.py]: Fixed

**Pull request merged**